pkg/kube/pod_controller.go

```diff
@@ -30,3 +45,139 @@ type PodController interface {
 	GetCommandExecutor() (PodCommandExecutor, error)
 	GetFileWriter() (PodFileWriter, error)
 }
+
+// podControllerProcessor aids in unit testing
+type podControllerProcessor interface {
+	createPod(ctx context.Context, cli kubernetes.Interface, options *PodOptions) (*corev1.Pod, error)
+	waitForPodReady(ctx context.Context, podName string) error
+	deletePod(ctx context.Context, namespace string, podName string, opts metav1.DeleteOptions) error
+}
+
+// podController specifies Kubernetes Client and PodOptions needed for creating Pod
+type podController struct {
+	cli        kubernetes.Interface
+	podOptions *PodOptions
+
+	pod         *corev1.Pod
+	podReady    bool
+	podName     string
+	stopTimeout time.Duration
+
+	pcp podControllerProcessor
+}
+
+type PodControllerOption func(p *podController)
+
+// WithPodControllerProcessor provides mechanism for passing fake podControllerProcessor for testing purposes.
+func WithPodControllerProcessor(processor podControllerProcessor) PodControllerOption {
+	return func(p *podController) {
+		p.pcp = processor
+	}
+}
+
+// NewPodController returns a new PodController given Kubernetes Client and PodOptions
+func NewPodController(cli kubernetes.Interface, options *PodOptions, opts ...PodControllerOption) PodController {
+	r := &podController{
+		cli:        cli,
+		podOptions: options,
+	}
+
+	r.pcp = r
+
+	for _, opt := range opts {
+		opt(r)
+	}
+
+	return r
+}
+
+func (p *podController) PodName() string {
+	return p.podName
+}
+
+func (p *podController) Pod() *corev1.Pod {
+	return p.pod
+}
+
+// StartPod creates pod and in case of success, it stores pod name for further use.
+// stopTimeout is also stored and will be used when StopPod will be called
+func (p *podController) StartPod(ctx context.Context, stopTimeout time.Duration) error {
+	if p.podName != "" {
+		return errors.Wrap(ErrPodControllerPodAlreadyStarted, "Failed to create pod")
+	}
+
+	if p.cli == nil || p.podOptions == nil {
+		return errors.Wrap(ErrPodControllerNotInitialized, "Failed to create pod")
+	}
+
+	pod, err := p.pcp.createPod(ctx, p.cli, p.podOptions)
+	if err != nil {
+		log.WithError(err).Print("Failed to create pod", field.M{"PodName": p.podOptions.Name, "Namespace": p.podOptions.Namespace})
+		return errors.Wrap(err, "Failed to create pod")
+	}
+
+	p.pod = pod
+	p.podName = pod.Name
+	p.stopTimeout = stopTimeout
+
+	return nil
+}
+
+// WaitForPod waits for POD readiness.
+func (p *podController) WaitForPodReady(ctx context.Context) error {
+	if p.podName == "" {
+		return errors.Wrap(ErrPodControllerPodNotStarted, "Pod failed to become ready in time")
```

Comment:
"Pod failed to become ready in time" - should this be "Pod not started" or "Pod name cannot be empty"?

Reply:
I think in this particular case we can `return ErrPodControllerPodNotStarted ` without wrapping, because this is not an issue happened during waiting. This method should not be called at all when Start failed.